activerecord/lib/active_record/associations/belongs_to_association.rb

```diff
@@ -30,7 +30,7 @@ def replace(record)
       end
 
       def target=(record)
-        replace_keys(record)
+        replace_keys(record) if different_target?(record)
```

Comment:
Wouldn't making the condition `record.nil? || different_target?(record)` incorrectly replace the keys if `target` is nil and `record` is `nil`? 

It looks like `BelongsToPolymorphicAssociation` needs to be changed to something like this to work properly for the null and non-null record cases (assuming it's safer than having `BelongsToPolymorphicAssociation#klass` return `NilClass` for the nil case):

```ruby
def different_target?(record) 
  super || record && record.class != klass || record.nil? && klass
end 
```

Reply:
Originally assigning belongs_to on destroyed object raised frozen error if `target` is `nil` and `record` is `nil`. 9e72e8e3949919f038ab584e4a42cd0502b1bc07
Since #32338 preloader doesn't assign `nil` to `target`, `record.nil? || different_target?(record)` is enough to prevent replacing the keys from preloader.

Reply:
Great point. I've pushed a change with a nil check.